airbyte-cdk/python/unit_tests/sources/streams/http/test_http_client.py

```diff
@@ -0,0 +1,312 @@
+        ),
+    ],
+)
+def test_duplicate_request_params_are_deduped(deduplicate_query_params, url, params, expected_url):
+    http_client = test_http_client()
+
+    if expected_url is None:
+        with pytest.raises(ValueError):
+            http_client._create_prepared_request(http_method="get", url=url, dedupe_query_params=deduplicate_query_params, params=params)
+    else:
+        prepared_request = http_client._create_prepared_request(http_method="get", url=url, dedupe_query_params=deduplicate_query_params, params=params)
+        assert prepared_request.url == expected_url
+
+def test_create_prepared_response_given_given_both_json_and_data_raises_request_body_exception():
+    http_client = test_http_client()
+
+    with pytest.raises(RequestBodyException):
+        http_client._create_prepared_request(http_method="get", url="https://test_base_url.com/v1/endpoint", json={"test": "json"}, data={"test": "data"})
+
+@pytest.mark.parametrize(
+    "json, data",
+    [
+        ({"test": "json"}, None),
+        (None, {"test": "data"}),
+    ],
+)
+def test_create_prepared_response_given_either_json_or_data_returns_valid_request(json, data):
+    http_client = test_http_client()
+    prepared_request = http_client._create_prepared_request(http_method="get", url="https://test_base_url.com/v1/endpoint", json=json, data=data)
+    assert prepared_request
+
+
+def test_connection_pool():
+    http_client = HttpClient(stream_name="test", logger=MagicMock(), authenticator=TokenAuthenticator("test-token"))
+    assert http_client._session.adapters["https://"]._pool_connections == 20
+
+def test_valid_basic_send_request(mocker):
+    http_client = test_http_client()
+    prepared_request = http_client._create_prepared_request(http_method="get", url="https://test_base_url.com/v1/endpoint")
+    response = requests.Response()
+    response.status_code = 200
+    response._content = b'{"test": "response"}'
+
+    mocker.patch.object(http_client, "_create_prepared_request", return_value=prepared_request)
+    mocker.patch.object(requests.Session, "send", return_value=response)
+
+    returned_request, returned_response = http_client.send_request(http_method="get", url="https://test_base_url.com/v1/endpoint")
+
+    assert returned_request == prepared_request
+    assert returned_response == response
+
+def test_send_raises_airbyte_traced_exception_with_fail_response_action(mocker):
+    http_client = test_http_client()
+    prepared_request = http_client._create_prepared_request(http_method="get", url="https://test_base_url.com/v1/endpoint")
+    response = requests.Response()
+    response.status_code = 400
+
+    mocker.patch.object(requests.Session, "send", return_value=response)
+    mocker.patch.object(http_client._error_handler, "interpret_response", return_value=(ResponseAction.FAIL, FailureType.system_error, "test error message"))
+
+    with pytest.raises(AirbyteTracedException):
+        http_client._send(prepared_request, {})
```

Comment:
Should we validate that we didn't retry here on in another test?

Reply:
Yes, will add.